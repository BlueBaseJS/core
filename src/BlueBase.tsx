import { Analytics, Logger } from './api';
import {
	ComponentCollection,
	ComponentRegistry,
	ConfigCollection,
	ConfigRegistry,
	HookNestedCollection,
	HookRegistry,
	PluginCollection,
	PluginRegistry,
	ThemeCollection,
	ThemeRegistry,
} from './registries';
import { BlueBaseProvider } from './Context';
import React from 'react';
import { ThemeProvider } from './themes';
import { getComponent } from './getComponent';
import systemHooks from './hooks';

export interface BootOptions {
	components: ComponentCollection,
	configs: ConfigCollection,
	hooks: HookNestedCollection,
	plugins: PluginCollection,
	themes: ThemeCollection,
	// routes: Plugin[]
}

export class BlueBase {

	// APIs
	public Analytics = new Analytics(this);
	public Logger = new Logger(this);

	// Registries
	public Components = new ComponentRegistry(this);
	public Configs = new ConfigRegistry(this);
	public Hooks = new HookRegistry(this);
	public Plugins = new PluginRegistry(this);
	public Themes = new ThemeRegistry(this);

	// Flags
	public booted = false;

	private bootOptions: BootOptions = {
		components: {},
		configs: {},
		hooks: {},
		plugins: [],
		themes: [],
	};

	public async boot(options?: Partial<BootOptions> & { children?: React.ReactNode }) {

		this.bootOptions = { ...this.bootOptions, ...options };

		await this.Hooks.registerNestedCollection(systemHooks);
		await this.Hooks.registerNestedCollection(this.bootOptions.hooks);

		await this.Hooks.run('bluebase.boot', this.bootOptions);

		const SystemApp = getComponent('SystemApp');

		// Set View
		// const SystemApp = this.Components.resolve('SystemApp');
		// SystemApp = await this.Hooks.run('bluebase.system.app', SystemApp);

		const BlueBaseRoot = () => (
			<BlueBaseProvider value={this}>
				<ThemeProvider>
					<SystemApp>
						{options && options.children}
					</SystemApp>
				</ThemeProvider>
			</BlueBaseProvider>
		);

		this.booted = true;

		return BlueBaseRoot;
	}
}