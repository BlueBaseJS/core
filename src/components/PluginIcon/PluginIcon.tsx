import { BlueBase } from '../../BlueBase';
import { BlueBaseConsumer } from '../../Context';
import React from 'react';
import { resolveThunk } from '../../utils';

export interface PluginIconProps {
	slug: string;
	size?: number,
	[key: string]: any
}

export class PluginIcon extends React.PureComponent<PluginIconProps> {

	render() {

		return (
			<BlueBaseConsumer>
				{(BB: BlueBase) => {

					const { slug, ...rest } = this.props;

					const plugin = BB.Plugins.get(slug);

					if (!plugin) {
						throw new Error(`There's no pluign registered with "${slug}" key in the registry.`);
					}

					if (!plugin.icon) {
						return null;
					}

					const iconProps = resolveThunk(plugin.icon, plugin, BB);

					return <BB.Components.DynamicIcon {...iconProps} {...rest} />;
				}}
			</BlueBaseConsumer>
		);
	}
}
