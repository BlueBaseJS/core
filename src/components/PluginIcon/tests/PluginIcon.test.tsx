import * as Component from '../..';
import { BlueBase } from '../../../BlueBase';
import { PluginIcon } from '../PluginIcon';
import React from 'react';
import TestRenderer from 'react-test-renderer';

beforeEach(() => {
	jest.resetModules();
});

const mockContext = jest.fn();
const BB = new BlueBase();

jest.mock('../../../Context', () => ({
	BlueBaseConsumer: ({ children }: { children: any }) => children(BB)
}));

describe('PluginIcon', () => {
	beforeEach(() => {
		mockContext.mockReset();
	});
	it(`Snapshot PluginIcon component`, async () => {
		await BB.Components.register('DynamicIcon', Component.DynamicIcon);
		await BB.Plugins.register({
			icon: {
				source: '',
				type: 'image',
			},
			name: 'DummyPlugin',
			slug: 'dummy-plugin',
		});
		const component = TestRenderer.create(
			<PluginIcon slug="dummy-plugin" />
		);
		const tree = component.toTree();
		expect(tree).toMatchSnapshot();
	});

});
