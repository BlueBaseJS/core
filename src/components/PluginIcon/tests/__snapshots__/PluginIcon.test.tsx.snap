// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginIcon Snapshot PluginIcon component 1`] = `
Object {
  "instance": PluginIcon {
    "_reactInternalFiber": FiberNode {
      "_debugID": 3,
      "_debugIsCurrentlyTiming": false,
      "_debugOwner": null,
      "_debugSource": null,
      "alternate": null,
      "child": FiberNode {
        "_debugID": 4,
        "_debugIsCurrentlyTiming": false,
        "_debugOwner": [Circular],
        "_debugSource": null,
        "alternate": null,
        "child": FiberNode {
          "_debugID": 5,
          "_debugIsCurrentlyTiming": false,
          "_debugOwner": [Circular],
          "_debugSource": null,
          "alternate": null,
          "child": FiberNode {
            "_debugID": 6,
            "_debugIsCurrentlyTiming": false,
            "_debugOwner": [Circular],
            "_debugSource": null,
            "alternate": null,
            "child": null,
            "childExpirationTime": 0,
            "effectTag": 1,
            "elementType": [Function],
            "expirationTime": 0,
            "firstContextDependency": null,
            "firstEffect": null,
            "index": 0,
            "key": null,
            "lastEffect": null,
            "memoizedProps": Object {
              "children": [Function],
            },
            "memoizedState": null,
            "mode": 0,
            "nextEffect": null,
            "pendingProps": Object {
              "children": [Function],
            },
            "ref": null,
            "return": [Circular],
            "sibling": null,
            "stateNode": null,
            "tag": 0,
            "type": [Function],
            "updateQueue": null,
          },
          "childExpirationTime": 0,
          "effectTag": 1,
          "elementType": [Function],
          "expirationTime": 0,
          "firstContextDependency": null,
          "firstEffect": null,
          "index": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": Object {
            "size": 100,
            "source": "",
            "type": "image",
          },
          "memoizedState": null,
          "mode": 0,
          "nextEffect": null,
          "pendingProps": Object {
            "size": 100,
            "source": "",
            "type": "image",
          },
          "ref": null,
          "return": [Circular],
          "sibling": null,
          "stateNode": null,
          "tag": 0,
          "type": [Function],
          "updateQueue": null,
        },
        "childExpirationTime": 0,
        "effectTag": 1,
        "elementType": [Function],
        "expirationTime": 0,
        "firstContextDependency": null,
        "firstEffect": null,
        "index": 0,
        "key": null,
        "lastEffect": null,
        "memoizedProps": Object {
          "children": [Function],
        },
        "memoizedState": null,
        "mode": 0,
        "nextEffect": null,
        "pendingProps": Object {
          "children": [Function],
        },
        "ref": null,
        "return": [Circular],
        "sibling": null,
        "stateNode": null,
        "tag": 0,
        "type": [Function],
        "updateQueue": null,
      },
      "childExpirationTime": 0,
      "effectTag": 1,
      "elementType": [Function],
      "expirationTime": 0,
      "firstContextDependency": null,
      "firstEffect": null,
      "index": 0,
      "key": null,
      "lastEffect": null,
      "memoizedProps": Object {
        "slug": "dummy-plugin",
      },
      "memoizedState": null,
      "mode": 0,
      "nextEffect": null,
      "pendingProps": Object {
        "slug": "dummy-plugin",
      },
      "ref": null,
      "return": FiberNode {
        "_debugID": 1,
        "_debugIsCurrentlyTiming": false,
        "_debugOwner": null,
        "_debugSource": null,
        "alternate": FiberNode {
          "_debugID": 1,
          "_debugIsCurrentlyTiming": false,
          "_debugOwner": null,
          "_debugSource": null,
          "alternate": [Circular],
          "child": null,
          "childExpirationTime": 0,
          "effectTag": 0,
          "elementType": null,
          "expirationTime": 1073741823,
          "firstContextDependency": null,
          "firstEffect": null,
          "index": 0,
          "key": null,
          "lastEffect": null,
          "memoizedProps": null,
          "memoizedState": null,
          "mode": 0,
          "nextEffect": null,
          "pendingProps": null,
          "ref": null,
          "return": null,
          "sibling": null,
          "stateNode": Object {
            "containerInfo": Object {
              "children": Array [],
              "createNodeMock": [Function],
              "tag": "CONTAINER",
            },
            "context": Object {},
            "current": [Circular],
            "didError": false,
            "earliestPendingTime": 0,
            "earliestSuspendedTime": 0,
            "expirationTime": 0,
            "finishedWork": null,
            "firstBatch": null,
            "hydrate": false,
            "latestPendingTime": 0,
            "latestPingedTime": 0,
            "latestSuspendedTime": 0,
            "nextExpirationTimeToWorkOn": 0,
            "nextScheduledRoot": null,
            "pendingChildren": null,
            "pendingCommitExpirationTime": 0,
            "pendingContext": null,
            "timeoutHandle": -1,
          },
          "tag": 3,
          "type": null,
          "updateQueue": Object {
            "baseState": null,
            "firstCapturedEffect": null,
            "firstCapturedUpdate": null,
            "firstEffect": null,
            "firstUpdate": Object {
              "callback": null,
              "expirationTime": 1073741823,
              "next": null,
              "nextEffect": null,
              "payload": Object {
                "element": <PluginIcon
                  slug="dummy-plugin"
                />,
              },
              "tag": 0,
            },
            "lastCapturedEffect": null,
            "lastCapturedUpdate": null,
            "lastEffect": null,
            "lastUpdate": Object {
              "callback": null,
              "expirationTime": 1073741823,
              "next": null,
              "nextEffect": null,
              "payload": Object {
                "element": <PluginIcon
                  slug="dummy-plugin"
                />,
              },
              "tag": 0,
            },
          },
        },
        "child": [Circular],
        "childExpirationTime": 0,
        "effectTag": 0,
        "elementType": null,
        "expirationTime": 0,
        "firstContextDependency": null,
        "firstEffect": [Circular],
        "index": 0,
        "key": null,
        "lastEffect": [Circular],
        "memoizedProps": null,
        "memoizedState": Object {
          "element": <PluginIcon
            slug="dummy-plugin"
          />,
        },
        "mode": 0,
        "nextEffect": null,
        "pendingProps": null,
        "ref": null,
        "return": null,
        "sibling": null,
        "stateNode": Object {
          "containerInfo": Object {
            "children": Array [],
            "createNodeMock": [Function],
            "tag": "CONTAINER",
          },
          "context": Object {},
          "current": [Circular],
          "didError": false,
          "earliestPendingTime": 0,
          "earliestSuspendedTime": 0,
          "expirationTime": 0,
          "finishedWork": null,
          "firstBatch": null,
          "hydrate": false,
          "latestPendingTime": 0,
          "latestPingedTime": 0,
          "latestSuspendedTime": 0,
          "nextExpirationTimeToWorkOn": 0,
          "nextScheduledRoot": null,
          "pendingChildren": null,
          "pendingCommitExpirationTime": 0,
          "pendingContext": null,
          "timeoutHandle": -1,
        },
        "tag": 3,
        "type": null,
        "updateQueue": Object {
          "baseState": Object {
            "element": <PluginIcon
              slug="dummy-plugin"
            />,
          },
          "firstCapturedEffect": null,
          "firstCapturedUpdate": null,
          "firstEffect": null,
          "firstUpdate": null,
          "lastCapturedEffect": null,
          "lastCapturedUpdate": null,
          "lastEffect": null,
          "lastUpdate": null,
        },
      },
      "sibling": null,
      "stateNode": [Circular],
      "tag": 1,
      "type": [Function],
      "updateQueue": null,
    },
    "_reactInternalInstance": Object {},
    "context": Object {},
    "props": Object {
      "slug": "dummy-plugin",
    },
    "refs": Object {},
    "state": null,
    "updater": Object {
      "enqueueForceUpdate": [Function],
      "enqueueReplaceState": [Function],
      "enqueueSetState": [Function],
      "isMounted": [Function],
    },
  },
  "nodeType": "component",
  "props": Object {
    "slug": "dummy-plugin",
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "component",
    "props": Object {
      "children": [Function],
    },
    "rendered": Object {
      "instance": null,
      "nodeType": "component",
      "props": Object {
        "size": 100,
        "source": "",
        "type": "image",
      },
      "rendered": Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "children": [Function],
        },
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  "type": [Function],
}
`;
